Source/CarthageKit/Project.swift

```diff
@@ -684,50 +698,85 @@ public final class Project {
-				} catch {
-					return false
-				}
-			}
-			.attemptMap { name -> Result<(), CarthageError> in
-				let dependencyCheckoutURL = dependencyCheckoutsURL.appendingPathComponent(name)
-				let subdirectoryPath = (CarthageProjectCheckoutsPath as NSString).appendingPathComponent(name)
-				let linkDestinationPath = relativeLinkDestinationForDependencyProject(dependency, subdirectory: subdirectoryPath)
-				do {
-					try fileManager.createSymbolicLink(atPath: dependencyCheckoutURL.carthage_path, withDestinationPath: linkDestinationPath)
-				} catch let error as NSError {
-					if !(error.domain == NSCocoaErrorDomain && error.code == NSFileWriteFileExistsError) {
-						return .failure(.writeFailed(dependencyCheckoutURL, error))
+		return self.dependencyProjects(for: dependency)
+			.zip(with: // file system objects which might conflict with symlinks
+				fileSystemObjects(for: dependency, withRepository: repositoryURL, at: CarthageProjectCheckoutsPath, flatMapTransform: { path in
+					let componentsRelativeToDirectoryURL = {
+						return URL(string: $0, relativeTo: self.directoryURL)?.standardizedFileURL.carthage_pathComponents
 					}
-				}
-				return .success()
-			}
+					if
+						let components = componentsRelativeToDirectoryURL(path),
+						let comparator = componentsRelativeToDirectoryURL(CarthageProjectCheckoutsPath),
+						Array(components.dropLast(1)) == comparator
+					{
+						return components.last!
+					} else {
+						return nil
+					}
+				})
+			)
+			.attemptMap { (dependencies: Set<ProjectIdentifier>, components: [String]) -> Result<(), CarthageError> in
+				let names = dependencies.filter { dependency in
+					// Filter out dependencies with names matching (case-insensitively) file system objects from git in `CarthageProjectCheckoutsPath`.
+					// Edge case warning on file system case-sensitivity. If a differently-cased file system object exists in git
+					// and is stored on a case-sensitive file system (like the Sierra preview of APFS), we currently preempt
+					// the non-conflicting symlink. Probably, nobody actually desires or needs the opposite behavior.
+					components.filter {
+						dependency.name.caseInsensitiveCompare($0) == .orderedSame
+					}.isEmpty
+				}.map { $0.name }
+
+				// If no `CarthageProjectCheckoutsPath`-housed symlinks are needed,
+				// return early after potentially adding submodules
+				// (which could be outside `CarthageProjectCheckoutsPath`).
+				if names.isEmpty { return .success() }
 
-		return SignalProducer<(), CarthageError>
-			.attempt {
 				do {
 					try fileManager.createDirectory(at: dependencyCheckoutsURL, withIntermediateDirectories: true)
 				} catch let error as NSError {
 					if !(error.domain == NSCocoaErrorDomain && error.code == NSFileWriteFileExistsError) {
 						return .failure(.writeFailed(dependencyCheckoutsURL, error))
 					}
 				}
+
+				for name in names {
```

Comment:
I don't think there's an easy way to do that. But we could do the check for each name.

It could also work to add a function that works like `mkdir -p`.